### DISCUSSION TITLE
Considering Dates as a Data Type in EvalEx

### DISCUSSION THREAD

**Discussion by vincentlejeune** [ezylang/EvalEx#379] [Ideas]
I've been able to work with expressions that involve numbers, strings, lists, maps, and dates. However, I'm aware that dates are not currently supported as a standard data type in EvalEx. I was wondering if it would be worth considering adding support for dates in EvalEx?

**Comment by uklimaschewski** [comment]
Hi,

yes, date types were also on my mind, when I designed version 3 of EvalEx.
My idea was to go with at least 2 new data types:
* _DATE_TIME_: Defines a date and time in years, months, days, hours, minutes, seconds and milliseconds.
* _PERIOD_: Defines an amount of time in years, months, days, hours, minutes, seconds and milliseconds.

Some basic operators could be modified to support these types:
* _DATE_TIME_ + _PERIOD_ will result in a new _DATE_TIME_, adding the amount of date/time defined in the period.
* _PERIOD_ + _PERIOD_ will add two periods.
* _DATE_TIME_ - _PERIOD_ will result in a new _DATE_TIME_, subtracting the amount of date/time defined in the period.
* _DATE_TIME_ - _DATE_TIME_ will result in a _PERIOD_ that is the difference between two dates, negative values are possible.
* _PERIOD_ - _PERIOD_ will subtract two periods.

Also some date functions could be added:
* dt_parse(string date, string format)
* dt_format(date_time date, string format)
* dt_to_millis(date_time or period)
* dt_date_from_millis(number)
* dt_period_from_millis(number)

Support for time zones is an open question for me.

I think that would be a very good additional feature for e.g. version 3.1 of EvalEx.

**Reply by vincentlejeune** [reply]
Hi @uklimaschewski,

Thank you for your response.

I agree that adding DATE_TIME and PERIOD as new data types would be a great addition to EvalEx. However, I was wondering if it would also be possible to include a DATE type that defines a date in years, months, and days.

I also agree that time zones are an important consideration. 

It would be useful to have methods to easily extract date and time components, such as year, month, day, hour, minute, and second.

Do you have any suggestions on the syntax to use for writing a DATE_TIME and PERIOD in an expression?

**Comment by uklimaschewski** [comment]
I haven't thought this all in detail, so this are just some ideas.

Writing _DATE_TIME_ and _PERIOD_ in an expression could be achieved easiest by using a function, some kind of constructor, e.g.:

`dt_date_time(2023, 5, 1, 14, 30, 21) - dt_period_days(1)` 

The `dt_date_time` parameters could be implemented using variable number of arguments, so you can e.g. specify the date part only.
I don't really see the need to have date literals in an expression, I think it is a very rare use case to have a fixed date in an expression.

Accessing the _DATE_TIME_ and _PERIOD_ parts could also be done through functions, or more elegant by using the structure operator:
`my_date.year + "/" + my_date.month`

Also, I think a _DATE_ could be seen as a _DATE_TIME_ with the time fields set to zero.

**Reply by vincentlejeune** [reply]
Thank you for your suggestions, @uklimaschewski .

Regarding the syntax for writing DATE_TIME and PERIOD in an expression, I understand your proposal to use a constructor function, such as 'dt_date_time' and 'dt_period_days'. However, I would like to point out that there may be cases where having a fixed date in an expression is useful, such as when working with specific events or deadlines.

In addition, I'd like to mention that a DATE may not necessarily be equivalent to a DATE_TIME with time fields set to zero. For example, a person's birthday may be considered a date, but may not have the same significance as a specific point in time.

That being said, I appreciate your response and would like to ask if you are planning to work on implementing these features in the near future, or if you think contributions would be welcome.
